sklearn/preprocessing/_polynomial.py

```diff
@@ -343,29 +421,52 @@ def transform(self, X):
         )
 
         n_samples, n_features = X.shape
-
+        max_int32 = np.iinfo(np.int32).max
         if sparse.isspmatrix_csr(X):
             if self._max_degree > 3:
                 return self.transform(X.tocsc()).tocsr()
             to_stack = []
             if self.include_bias:
                 to_stack.append(
-                    sparse.csc_matrix(np.ones(shape=(n_samples, 1), dtype=X.dtype))
+                    sparse.csr_matrix(np.ones(shape=(n_samples, 1), dtype=X.dtype))
                 )
             if self._min_degree <= 1 and self._max_degree > 0:
                 to_stack.append(X)
+
+            cumulative_size = sum(mat.shape[1] for mat in to_stack)
             for deg in range(max(2, self._min_degree), self._max_degree + 1):
-                Xp_next = _csr_polynomial_expansion(
-                    X.data, X.indices, X.indptr, X.shape[1], self.interaction_only, deg
+                expanded = _create_expansion(
+                    X=X,
+                    interaction_only=self.interaction_only,
+                    deg=deg,
+                    n_features=n_features,
+                    cumulative_size=cumulative_size,
                 )
-                if Xp_next is None:
-                    break
-                to_stack.append(Xp_next)
+                if expanded is not None:
+                    to_stack.append(expanded)
+                    cumulative_size += expanded.shape[1]
             if len(to_stack) == 0:
                 # edge case: deal with empty matrix
                 XP = sparse.csr_matrix((n_samples, 0), dtype=X.dtype)
             else:
-                XP = sparse.hstack(to_stack, format="csr")
+                # `scipy.sparse.hstack` breaks in scipy<1.9.2
+                # when `n_output_features_ > max_int32`
+                all_int32 = all(mat.indices.dtype == np.int32 for mat in to_stack)
+                if (
+                    sp_version < parse_version("1.9.2")
+                    and self.n_output_features_ > max_int32
+                    and all_int32
+                ):
+                    raise ValueError(
```

Comment:
I think we can ignore the lack of test coverage for the `< 1.9.2` bug, unless you find a way to trigger this case without changing the CI config.

Reply:
@thomasjpfan and I discussed the possibility of monkeypatching the scipy version _down_ on CI w/ scipy `>1.9.2` to trigger the `ValueError`. Do you think it would be simpler to ignore the lack of coverage, or is it worth including the monkeypatch in what is already a fairly complex test?

Reply:
No opinion. Do as you wish.